File: lib/data/store.dart
import 'dart:io';

import 'package:file_saver/file_saver.dart';
import 'package:flutter/foundation.dart';

import 'models.dart';
import 'vault.dart';

class ImportSummary {
  const ImportSummary({
    required this.hostsAdded,
    required this.hostsUpdated,
    required this.identitiesAdded,
    required this.identitiesUpdated,
    required this.groupsAdded,
  });

  final int hostsAdded;
  final int hostsUpdated;
  final int identitiesAdded;
  final int identitiesUpdated;
  final int groupsAdded;

  int get total =>
      hostsAdded +
      hostsUpdated +
      identitiesAdded +
      identitiesUpdated +
      groupsAdded;
}

class VaultStore extends ChangeNotifier {
  VaultStore._();

  static final VaultStore instance = VaultStore._();

  Vault? _vault;
  VaultData _data = const VaultData();

  bool get isUnlocked => _vault != null;

  List<Host> get hosts => List.unmodifiable(_data.hosts);
  List<Identity> get identities => List.unmodifiable(_data.identities);
  List<HostGroup> get groups => List.unmodifiable(_data.groups);

  static Future<bool> vaultExists() => Vault.exists();

  Future<void> create(String password) async {
    _vault = await Vault.create(password: password);
    _data = const VaultData();
    notifyListeners();
  }

  Future<void> unlock(String password) async {
    final vault = await Vault.open(password: password);
    _data = await vault.read();
    _vault = vault;
    notifyListeners();
  }

  void lock() {
    _vault = null;
    _data = const VaultData();
    notifyListeners();
  }

  Future<void> changeMasterPassword(String newPassword) async {
    await _requireVault().changePassword(newPassword);
  }

  Identity? identityFor(Host host) {
    final inline = host.inlineIdentity;
    if (inline != null) return inline;
    final id = host.identityId;
    if (id == null) return null;
    for (final identity in _data.identities) {
      if (identity.id == id) return identity;
    }
    return null;
  }

  Identity? identityById(String? id) {
    if (id == null) return null;
    for (final identity in _data.identities) {
      if (identity.id == id) return identity;
    }
    return null;
  }

  HostGroup? groupById(String? id) {
    if (id == null) return null;
    for (final group in _data.groups) {
      if (group.id == id) return group;
    }
    return null;
  }

  /// Ungrouped hosts come last under a null key.
  Map<HostGroup?, List<Host>> hostsByGroup() {
    final buckets = <HostGroup?, List<Host>>{};
    for (final group in _data.groups) {
      buckets[group] = <Host>[];
    }
    final ungrouped = <Host>[];
    for (final host in _data.hosts) {
      final group = groupById(host.groupId);
      if (group == null) {
        ungrouped.add(host);
      } else {
        buckets.putIfAbsent(group, () => <Host>[]).add(host);
      }
    }
    buckets.removeWhere((_, hosts) => hosts.isEmpty);
    if (ungrouped.isNotEmpty) buckets[null] = ungrouped;
    return buckets;
  }

  Future<void> saveHost(Host host) => _mutate(() {
    final hosts = [..._data.hosts];
    final index = hosts.indexWhere((h) => h.id == host.id);
    if (index >= 0) {
      hosts[index] = host;
    } else {
      hosts.add(host);
    }
    _data = VaultData(
      hosts: hosts,
      identities: _data.identities,
      groups: _data.groups,
    );
  });

  Future<void> deleteHost(String id) => _mutate(() {
    _data = VaultData(
      hosts: _data.hosts.where((h) => h.id != id).toList(),
      identities: _data.identities,
      groups: _data.groups,
    );
  });

  Future<void> saveIdentity(Identity identity) => _mutate(() {
    final identities = [..._data.identities];
    final index = identities.indexWhere((i) => i.id == identity.id);
    if (index >= 0) {
      identities[index] = identity;
    } else {
      identities.add(identity);
    }
    _data = VaultData(
      hosts: _data.hosts,
      identities: identities,
      groups: _data.groups,
    );
  });

  /// Detaches the identity from every host that used it, so no host is left
  /// pointing at an id that no longer resolves.
  Future<void> deleteIdentity(String id) => _mutate(() {
    _data = VaultData(
      hosts: _data.hosts
          .map((h) => h.identityId == id ? h.copyWith(clearIdentity: true) : h)
          .toList(),
      identities: _data.identities.where((i) => i.id != id).toList(),
      groups: _data.groups,
    );
  });

  Future<void> saveGroup(HostGroup group) => _mutate(() {
    final groups = [..._data.groups];
    final index = groups.indexWhere((g) => g.id == group.id);
    if (index >= 0) {
      groups[index] = group;
    } else {
      groups.add(group);
    }
    _data = VaultData(
      hosts: _data.hosts,
      identities: _data.identities,
      groups: groups,
    );
  });

  /// Hosts in the group survive; they just become ungrouped.
  Future<void> deleteGroup(String id) => _mutate(() {
    _data = VaultData(
      hosts: _data.hosts
          .map((h) => h.groupId == id ? h.copyWith(clearGroup: true) : h)
          .toList(),
      identities: _data.identities,
      groups: _data.groups.where((g) => g.id != id).toList(),
    );
  });

  Future<String?> exportVault({String fileName = 'ohmyssh'}) async {
    final bytes = await _requireVault().file.readAsBytes();
    return FileSaver.instance.saveAs(
      name: fileName,
      bytes: Uint8List.fromList(bytes),
      fileExtension: 'vault',
      mimeType: MimeType.json,
    );
  }

  /// Merges another vault into this one: matching ids are overwritten, new ids
  /// appended.
  Future<ImportSummary> importVault({
    required File file,
    required String password,
  }) async {
    final incoming = await (await Vault.open(
      password: password,
      file: file,
    )).read();

    var hostsAdded = 0, hostsUpdated = 0;
    var identitiesAdded = 0, identitiesUpdated = 0;
    var groupsAdded = 0;

    final hosts = [..._data.hosts];
    for (final host in incoming.hosts) {
      final index = hosts.indexWhere((h) => h.id == host.id);
      if (index >= 0) {
        hosts[index] = host;
        hostsUpdated++;
      } else {
        hosts.add(host);
        hostsAdded++;
      }
    }

    final identities = [..._data.identities];
    for (final identity in incoming.identities) {
      final index = identities.indexWhere((i) => i.id == identity.id);
      if (index >= 0) {
        identities[index] = identity;
        identitiesUpdated++;
      } else {
        identities.add(identity);
        identitiesAdded++;
      }
    }

    final groups = [..._data.groups];
    for (final group in incoming.groups) {
      final index = groups.indexWhere((g) => g.id == group.id);
      if (index >= 0) {
        groups[index] = group;
      } else {
        groups.add(group);
        groupsAdded++;
      }
    }

    await _mutate(() {
      _data = VaultData(hosts: hosts, identities: identities, groups: groups);
    });

    return ImportSummary(
      hostsAdded: hostsAdded,
      hostsUpdated: hostsUpdated,
      identitiesAdded: identitiesAdded,
      identitiesUpdated: identitiesUpdated,
      groupsAdded: groupsAdded,
    );
  }

  Vault _requireVault() {
    final vault = _vault;
    if (vault == null) throw const VaultException('Vault is locked');
    return vault;
  }

  Future<void> _mutate(void Function() change) async {
    final vault = _requireVault();
    change();
    await vault.save(_data);
    notifyListeners();
  }
}

File: shared/src/commonMain/kotlin/com/example/ohmyssh/data/Store.kt
package com.example.ohmyssh.data

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.example.ohmyssh.platform.AppFiles
import com.example.ohmyssh.platform.FilePick

class ImportSummary(
    val hostsAdded: Int,
    val hostsUpdated: Int,
    val identitiesAdded: Int,
    val identitiesUpdated: Int,
    val groupsAdded: Int,
    val serialDevicesAdded: Int,
) {
    val total: Int
        get() = hostsAdded + hostsUpdated + identitiesAdded + identitiesUpdated +
            groupsAdded + serialDevicesAdded
}

object VaultStore {
    private var vault: Vault? = null

    var data: VaultData by mutableStateOf(VaultData())
        private set

    var isUnlocked: Boolean by mutableStateOf(false)
        private set

    val hosts: List<Host> get() = data.hosts
    val identities: List<Identity> get() = data.identities
    val groups: List<HostGroup> get() = data.groups
    val serialDevices: List<SerialDevice> get() = data.serialDevices

    fun vaultExists(): Boolean = Vault.exists()

    suspend fun create(password: String) {
        val created = Vault.create(password)
        vault = created
        data = VaultData()
        HistoryStore.open(created)
        isUnlocked = true
    }

    suspend fun unlock(password: String) {
        val opened = Vault.open(password)
        data = opened.read()
        vault = opened
        HistoryStore.open(opened)
        isUnlocked = true
    }

    fun lock() {
        vault = null
        data = VaultData()
        HistoryStore.close()
        isUnlocked = false
    }

    fun deleteVault() {
        val target = requireVault()
        // The history goes first, while its key still exists to name the file.
        HistoryStore.wipe()
        if (AppFiles.exists(target.path)) AppFiles.delete(target.path)
        lock()
    }

    suspend fun verifyPassword(password: String): Boolean = try {
        Vault.open(password, requireVault().path)
        true
    } catch (_: WrongPasswordException) {
        false
    }

    suspend fun changeMasterPassword(newPassword: String) {
        // The history is encrypted under the same password and has to travel
        // with it, or the new password would open the vault and nothing else.
        requireVault().changePassword(
            newPassword,
            sidecars = listOf(kHistoryFileName to kHistoryFormat),
        )
        if (AutoLogin.isEnabled()) AutoLogin.enable(newPassword)
    }

    fun identityFor(host: Host): Identity? {
        host.inlineIdentity?.let { return it }
        val id = host.identityId ?: return null
        return data.identities.firstOrNull { it.id == id }
    }

    fun identityById(id: String?): Identity? =
        id?.let { wanted -> data.identities.firstOrNull { it.id == wanted } }

    fun groupById(id: String?): HostGroup? =
        id?.let { wanted -> data.groups.firstOrNull { it.id == wanted } }

    fun hostsByGroup(): Map<HostGroup?, List<Host>> {
        val buckets = LinkedHashMap<HostGroup?, MutableList<Host>>()
        for (group in data.groups) buckets[group] = mutableListOf()
        val ungrouped = mutableListOf<Host>()
        for (host in data.hosts) {
            val group = groupById(host.groupId)
            if (group == null) ungrouped.add(host) else buckets.getValue(group).add(host)
        }
        buckets.entries.removeAll { it.value.isEmpty() }
        if (ungrouped.isNotEmpty()) buckets[null] = ungrouped
        return buckets
    }

    suspend fun saveHost(host: Host) = mutate {
        data = data.copy(hosts = upsert(data.hosts, host) { it.id })
    }

    suspend fun deleteHost(id: String) = mutate {
        data = data.copy(hosts = data.hosts.filter { it.id != id })
    }

    suspend fun saveIdentity(identity: Identity) = mutate {
        data = data.copy(identities = upsert(data.identities, identity) { it.id })
    }

    suspend fun deleteIdentity(id: String) = mutate {
        data = data.copy(
            hosts = data.hosts.map { if (it.identityId == id) it.copy(identityId = null) else it },
            identities = data.identities.filter { it.id != id },
        )
    }

    suspend fun saveGroup(group: HostGroup) = mutate {
        data = data.copy(groups = upsert(data.groups, group) { it.id })
    }

    suspend fun deleteGroup(id: String) = mutate {
        data = data.copy(
            hosts = data.hosts.map { if (it.groupId == id) it.copy(groupId = null) else it },
            groups = data.groups.filter { it.id != id },
        )
    }

    suspend fun saveSerialDevice(device: SerialDevice) = mutate {
        data = data.copy(serialDevices = upsert(data.serialDevices, device) { it.id })
    }

    suspend fun deleteSerialDevice(id: String) = mutate {
        data = data.copy(serialDevices = data.serialDevices.filter { it.id != id })
    }

    private fun <T> upsert(items: List<T>, item: T, idOf: (T) -> String): List<T> {
        val next = items.toMutableList()
        val index = next.indexOfFirst { idOf(it) == idOf(item) }
        if (index >= 0) next[index] = item else next.add(item)
        return next
    }

    suspend fun exportVault(fileName: String = "ohmyssh"): String? {
        val bytes = requireVault().readFileBytes()
        return FilePick.saveFile(name = fileName, extension = "vault", bytes = bytes)
    }

    suspend fun importVault(fileText: String, password: String): ImportSummary {
        val incoming = Vault.decryptData(fileText, password)

        var hostsAdded = 0
        var hostsUpdated = 0
        var identitiesAdded = 0
        var identitiesUpdated = 0
        var groupsAdded = 0
        var serialDevicesAdded = 0

        val hosts = data.hosts.toMutableList()
        for (host in incoming.hosts) {
            val index = hosts.indexOfFirst { it.id == host.id }
            if (index >= 0) {
                hosts[index] = host
                hostsUpdated++
            } else {
                hosts.add(host)
                hostsAdded++
            }
        }

        val identities = data.identities.toMutableList()
        for (identity in incoming.identities) {
            val index = identities.indexOfFirst { it.id == identity.id }
            if (index >= 0) {
                identities[index] = identity
                identitiesUpdated++
            } else {
                identities.add(identity)
                identitiesAdded++
            }
        }

        val groups = data.groups.toMutableList()
        for (group in incoming.groups) {
            val index = groups.indexOfFirst { it.id == group.id }
            if (index >= 0) groups[index] = group else {
                groups.add(group)
                groupsAdded++
            }
        }

        val serialDevices = data.serialDevices.toMutableList()
        for (device in incoming.serialDevices) {
            val index = serialDevices.indexOfFirst { it.id == device.id }
            if (index >= 0) serialDevices[index] = device else {
                serialDevices.add(device)
                serialDevicesAdded++
            }
        }

        mutate {
            data = VaultData(
                hosts = hosts,
                identities = identities,
                groups = groups,
                serialDevices = serialDevices,
            )
        }

        return ImportSummary(
            hostsAdded = hostsAdded,
            hostsUpdated = hostsUpdated,
            identitiesAdded = identitiesAdded,
            identitiesUpdated = identitiesUpdated,
            groupsAdded = groupsAdded,
            serialDevicesAdded = serialDevicesAdded,
        )
    }

    private fun requireVault(): Vault =
        vault ?: throw VaultException("Vault is locked")

    private suspend fun mutate(change: () -> Unit) {
        val target = requireVault()
        change()
        target.save(data)
    }
}
